filters/encrypt/filter.go

```diff
@@ -0,0 +1,554 @@
+				return fmt.Errorf("%s: %w", op, err)
+			}
+		case HmacSha256Operation:
+			if data, err = ef.hmacSha256(ctx, raw, opt...); err != nil {
+				return fmt.Errorf("%s: %w", op, err)
+			}
+		case RedactOperation:
+			data = RedactedData
+		default: // catch UnknownOperation, NoOperation and everything else
+			return fmt.Errorf("%s: unknown filter operation for field: %s: %w", op, classificationTag.Operation, ErrInvalidParameter)
+		}
+		if opts.withPointerstructureInfo != nil {
+			if _, err := pointerstructure.Set(opts.withPointerstructureInfo.i, opts.withPointerstructureInfo.pointer, data); err != nil {
+				return fmt.Errorf("%s: %w", op, err)
+			}
+		} else {
+			if err := setValue(fv, data); err != nil {
+				return fmt.Errorf("%s: %w", op, err)
+			}
+		}
+	default:
+		if err := setValue(fv, RedactedData); err != nil {
+			return fmt.Errorf("%s: %w", op, err)
+		}
+	}
+	return nil
+}
+
+func (ef *Filter) encrypt(ctx context.Context, data []byte, opt ...Option) (string, error) {
+	const op = "event.(Filter).encrypt"
+	if data == nil {
+		return "", fmt.Errorf("%s: missing data: %w", op, ErrInvalidParameter)
+	}
+	ef.l.Lock()
+	defer ef.l.Unlock()
+	opts := getOpts(opt...)
+	if ef.Wrapper == nil && opts.withWrapper == nil {
+		return "", fmt.Errorf("%s: missing wrapper: %w", op, ErrInvalidParameter)
+	}
+	var w wrapping.Wrapper
+	switch {
+	case opts.withWrapper != nil:
+		w = opts.withWrapper
+	default:
+		w = ef.Wrapper
+	}
+	blobInfo, err := w.Encrypt(ctx, data, nil)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", op, err)
+	}
+	marshaledBlob, err := proto.Marshal(blobInfo)
+	if err != nil {
+		return "", fmt.Errorf("error marshaling encrypted blob: %w", err)
+	}
+	return "encrypted:" + base64.RawURLEncoding.EncodeToString(marshaledBlob), nil
+}
+
+func (ef *Filter) hmacSha256(ctx context.Context, data []byte, opt ...Option) (string, error) {
+	const op = "event.(Filter).hmacSha256"
+	if data == nil {
+		return "", fmt.Errorf("%s: missing data: %w", op, ErrInvalidParameter)
+	}
+	ef.l.Lock()
+	defer ef.l.Unlock()
+	opts := getOpts(opt...)
+	if ef.Wrapper == nil && opts.withWrapper == nil {
+		return "", fmt.Errorf("%s: missing wrapper: %w", op, ErrInvalidParameter)
+	}
+	var w wrapping.Wrapper
+	switch {
+	case opts.withWrapper != nil:
+		w = opts.withWrapper
+	default:
+		w = ef.Wrapper
+	}
+
+	var salt []byte
+	switch {
+	case opts.withSalt != nil:
+		salt = make([]byte, len(opts.withSalt))
+		copy(salt, opts.withSalt)
+	default:
+		salt = make([]byte, len(ef.HmacSalt))
+		copy(salt, ef.HmacSalt)
+	}
+
+	var info []byte
+	switch {
+	case opts.withInfo != nil:
+		info = make([]byte, len(opts.withInfo))
+		copy(info, opts.withInfo)
+	default:
+		info = make([]byte, len(ef.HmacInfo))
+		copy(info, ef.HmacInfo)
+	}
+
+	reader, err := NewDerivedReader(w, 32, salt, info)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", op, err)
+	}
+	key, _, err := ed25519.GenerateKey(reader)
```

Comment:
I think this step is unnecessary. You're not using the key for anything, so no real reason to spend the time generating it. You can just read bytes out of the reader directly via https://golang.org/pkg/io/#ReadFull or so.

Reply:
ty!